File: parking_lot_system/parking_lot/vehicle.py
from enum import Enum
from abc import ABC, abstractmethod

class VehicleType(Enum):
    CAR = "car"
    MOTORCYCLE = "motorcycle"
    TRUCK = "truck"

class Vehicle(ABC):
    def __init__(self, license_plate, vehicle_type):
        self.license_plate = license_plate
        self.vehicle_type = vehicle_type

class Car(Vehicle):
    def __init__(self, license_plate):
        super().__init__(license_plate, VehicleType.CAR)

class Motorcycle(Vehicle):
    def __init__(self, license_plate):
        super().__init__(license_plate, VehicleType.MOTORCYCLE)

class Truck(Vehicle):
    def __init__(self, license_plate):
        super().__init__(license_plate, VehicleType.TRUCK)

File: parking_lot_system_go/parkinglot/vehicle.go
package parkinglot

type VehicleType int

const (
	VehicleTypeCar VehicleType = iota
	VehicleTypeMotorcycle
	VehicleTypeTruck
)

type Vehicle interface {
	GetLicensePlate() string
	GetType() VehicleType
}

type BaseVehicle struct {
	LicensePlate string
	Type         VehicleType
}

func (v *BaseVehicle) GetLicensePlate() string {
	return v.LicensePlate
}

func (v *BaseVehicle) GetType() VehicleType {
	return v.Type
}

type Car struct {
	BaseVehicle
}

func NewCar(licensePlate string) *Car {
	return &Car{BaseVehicle{LicensePlate: licensePlate, Type: VehicleTypeCar}}
}

type Motorcycle struct {
	BaseVehicle
}

func NewMotorcycle(licensePlate string) *Motorcycle {
	return &Motorcycle{BaseVehicle{LicensePlate: licensePlate, Type: VehicleTypeMotorcycle}}
}

type Truck struct {
	BaseVehicle
}

func NewTruck(licensePlate string) *Truck {
	return &Truck{BaseVehicle{LicensePlate: licensePlate, Type: VehicleTypeTruck}}
}
